File: day1/day1.rb
BUILDING = File.read("day1.txt").split ''

def silver b
  b.inject(0){|floor, i| floor += i == "(" ? 1 : -1 }  
end

def gold b
  floor = 0
  b.map{|i| i == "(" ? 1 : -1}.each.with_index { |dir, index|
    floor += dir
    return index if floor == -1
  }
end

puts "Silver: #{silver(BUILDING)}"
puts "Gold: #{gold(BUILDING)}"

File: 2015/day1.ml
fun map f [] = []
  | map f (x::xs) = (f x) :: map f xs;

fun dir(#"(") = 1
  | dir(#")") = ~1
  | dir _ = 0;

fun solve building =
  let
    fun solve' (floor, []) = floor
      | solve' (floor, x::xs) = solve' (floor+x, xs)
  in
    solve' (0, building)
  end;
